src/mpi/coll/transports/generic/transport.c

```diff
@@ -0,0 +1,1243 @@
+    vtxp->nbargs.sendrecv.dt = dt;
+    vtxp->nbargs.sendrecv.dest = dest;
+    vtxp->nbargs.sendrecv.comm = comm;
+    /* second request is not used */
+    vtxp->mpid_req[1] = NULL;
+
+    MPL_DBG_MSG_FMT(MPIR_DBG_COLL,VERBOSE,(MPL_DBG_FDEST,"TSP(generic) : sched [%d] [send_accumulate]\n", vtx_id));
+    return vtx_id;
+}
+
+/* Transport function to schedule a recv task */
+int MPIR_COLL_GENERIC_recv(void *buf,
+                                int count,
+                                MPIR_COLL_GENERIC_dt_t dt,
+                                int source,
+                                int tag,
+                                MPIR_COLL_GENERIC_comm_t * comm,
+                                MPIR_COLL_GENERIC_sched_t * sched, int n_invtcs, int *invtcs)
+{
+    /* assign a new vertex */
+    MPIR_COLL_GENERIC_vtx_t *vtxp;
+    int vtx_id = MPIR_COLL_GENERIC_get_new_vtx(sched, &vtxp);
+
+    sched->tag = tag;
+    vtxp->kind = MPIR_COLL_GENERIC_KIND_RECV;
+    MPIR_COLL_GENERIC_init_vtx(sched, vtxp, vtx_id);
+    MPIR_COLL_GENERIC_add_vtx_dependencies(sched, vtx_id, n_invtcs, invtcs);
+
+    /* record the arguments */
+    vtxp->nbargs.sendrecv.buf = buf;
+    vtxp->nbargs.sendrecv.count = count;
+    vtxp->nbargs.sendrecv.dt = dt;
+    vtxp->nbargs.sendrecv.dest = source;
+    vtxp->nbargs.sendrecv.comm = comm;
+    vtxp->mpid_req[1] = NULL;
+
+    MPL_DBG_MSG_FMT(MPIR_DBG_COLL,VERBOSE,(MPL_DBG_FDEST,"TSP(generic) : sched [%d] [recv]\n", vtx_id));
+    return vtx_id;
+}
+
+
+/* Internal transport function to do the reduce in the recv_reduce call */
+MPL_STATIC_INLINE_PREFIX int MPIR_COLL_GENERIC_queryfcn(void *data, MPI_Status * status)
+{
+    MPIR_COLL_GENERIC_recv_reduce_arg_t *rr = (MPIR_COLL_GENERIC_recv_reduce_arg_t *) data;
+    /* Check if recv has completed (mpid_req[0]==NULL) and reduce has not already been done (!rr->done) */
+    if (rr->vtxp->mpid_req[0] == NULL && !rr->done) {
+        MPI_Datatype dt = rr->datatype;
+        MPI_Op op = rr->op;
+
+        if (rr->flags == -1 || rr->flags & MPIR_COLL_FLAG_REDUCE_L) {
+            MPL_DBG_MSG_FMT(MPIR_DBG_COLL,VERBOSE,(MPL_DBG_FDEST,"  --> GENERIC transport (recv_reduce L) complete to %p\n", rr->inoutbuf));
+
+            MPIR_Reduce_local(rr->inbuf, rr->inoutbuf, rr->count, dt, op);
+        } else {
+            MPL_DBG_MSG_FMT(MPIR_DBG_COLL,VERBOSE,(MPL_DBG_FDEST,"  --> GENERIC transport (recv_reduce R) complete to %p\n", rr->inoutbuf));
+
+            MPIR_Reduce_local(rr->inoutbuf, rr->inbuf, rr->count, dt, op);
+            MPIR_Localcopy(rr->inbuf, rr->count, dt, rr->inoutbuf, rr->count, dt);
+        }
+
+        MPIR_Grequest_complete_impl(rr->vtxp->mpid_req[1]);
+        rr->done = 1;   /* recv_reduce is complete now */
+    }
+
+    /* TODO: Check if this has to be done everytime */
+    status->MPI_SOURCE = MPI_UNDEFINED;
+    status->MPI_TAG = MPI_UNDEFINED;
+    MPI_Status_set_cancelled(status, 0);
+    MPI_Status_set_elements(status, MPI_BYTE, 0);
```

Comment:
Don't use `MPI_` interfaces inside the library. Just call `MPIR_Status_set_elements_x_impl(status, MPI_BYTE, 0);`.